### ISSUE TITLE
PSR12.Classes.AnonClassDeclaration.SpaceAfterKeyword fires on `new class($arg)` despite PSR-12 not requiring it

### ISSUE THREAD

**Issue by dereuromark** [PHPCSStandards/PHP_CodeSniffer#1413]
## Describe the bug

`PSR12.Classes.AnonClassDeclaration.SpaceAfterKeyword` reports `Expected 1 space after class keyword; 0 found` on anonymous classes written as:

```php
$foo = new class($arg) extends Bar {
    public function __construct($a) {}
};
```

…requiring the user to insert a space: `new class ($arg)`.

However, the PSR-12 specification does not mandate a space between the `class` keyword and the constructor argument list of an anonymous class. Section 8 (Anonymous Classes) of PSR-12 references the rules for closures and shows anonymous class examples without touching on `class(` spacing specifically. The "one space after class keyword" rule is a PSR-2 rule written for named class declarations (`class Foo`), which pre-dates PHP 7's anonymous classes.

## Root cause

`PSR12\Sniffs\Classes\AnonClassDeclarationSniff` extends `PSR2\Sniffs\Classes\ClassDeclarationSniff` and unconditionally calls the parent's `processOpen()`:

https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/src/Standards/PSR12/Sniffs/Classes/AnonClassDeclarationSniff.php#L67

The parent's `processOpen()` in `PSR2\ClassDeclarationSniff` emits the `SpaceAfterKeyword` error here:

https://github.com/PHPCSStandards/PHP_CodeSniffer/blob/master/src/Standards/PSR2/Sniffs/Classes/ClassDeclarationSniff.php#L169

That check was designed for named class/interface/trait declarations where the next token is a class name (e.g. `class Foo`), and firing it on `new class(` appears to be an unintended side effect of the inheritance rather than a deliberate PSR-12 rule.

## Conflict with PHPCSExtra

This also makes `PSR12.Classes.AnonClassDeclaration.SpaceAfterKeyword` directly conflict with [`Universal.WhiteSpace.AnonClassKeywordSpacing`](https://github.com/PHPCSStandards/PHPCSExtra/blob/stable/Universal/Docs/WhiteSpace/AnonClassKeywordSpacingStandard.xml) from PHPCSExtra, whose default is no space (`new class(\$arg)`). Rulesets that include both standards currently have to `exclude` one of them.

## Expected behavior

The `PSR12` anon-class sniff should not enforce a space between `class` and `(` for anonymous classes with a constructor argument list, since PSR-12 does not require it.

Possible fixes:

1. Override `processOpen()` in `AnonClassDeclarationSniff` to skip the `SpaceAfterKeyword` check when the next non-whitespace token after `class` is `(`.
2. Or make the check in `PSR2\ClassDeclarationSniff::processOpen()` skip anonymous classes when called via the anon subclass.

## Versions

- PHP_CodeSniffer: current master / recent releases
- Standard: `PSR12`

## Reproduce

```php
<?php
$foo = new class('x') {
    public function __construct(string \$a) {}
};
```

Run with `--standard=PSR12`. The sniff reports `SpaceAfterKeyword` at the `class` keyword.

See https://github.com/php-collective/djot-php/pull/177/changes/c54dcc7691c2ad0bdb881bc9fe7b4abeb7017b32
and my current workaround required.

**Comment by jrfnl** [comment]
@dereuromark Please use the bug report template and don't use AI when submitting bug reports.

**Comment by jrfnl** [comment]
Note: the bug description contains multiple hallucinations and incorrect conclusions. This is not actionable or even worth having a discussion about as all the bullsh*t would need to be dispelled first.

I'm going to close this report now.

**Comment by dereuromark** [comment]
The actual issue is still true, the conflict is real as can be seen by the CI linked.
Answering like this is not really acceptable IMO.
Anyway, I will move on.

**Comment by jrfnl** [comment]
@dereuromark No offense meant, but as I stated, it would take me way too much time to actually write up and dispel all the things which are wrong in the issue description. Sorry, but that's just wasting my time.

I find it incredibly disrespectful if you don't even review whatever an AI vomits out before pushing it on maintainers.

I agree this is not an acceptable way to communicate. Please write your bug reports yourself and write them in a way that is not vomit and I will gladly have a conversation with you about them.

**Comment by dereuromark** [comment]
I looked over it, the code and analysis of the issue (conflict) is 100% correct and real.
I dont know what part was allegedly hallucinated, it wasnt the facts here. 

Claude appears to have mixed true observations with a shaky legal-style conclusion about what PSR-12 requires.
But does that invalidate anything here? I dont think so. In fact, that might only mean that the Universal sniff is the one that needs to change, if we want to follow PSR12 here more closely.

**Comment by jrfnl** [comment]
@dereuromark Again no. Short version:

PSR12 explicitly states - and I quote:
> Anonymous Classes MUST follow the same guidelines and principles as closures in the above section.
Src: https://www.php-fig.org/psr/psr-12/#8-anonymous-classes

> Closures MUST be declared with a space after the function keyword, and a space before and after the use keyword.
Src: https://www.php-fig.org/psr/psr-12/#7-closures

This combined leads to the "space after class keyword" for anonymous classes in the PSR12 ruleset being correct and not up for discussion.

The fact that that conflicts with a different sniff, which has its own specification in another repo, is neither here nor there.

If you use both sniffs and that causes a conflict, that is a configuration issue for your own ruleset and not a sniff issue.

Aside from that, the [Universal sniff is configurable](https://github.com/PHPCSStandards/PHPCSExtra?tab=readme-ov-file#universalwhitespaceanonclasskeywordspacing-wrench-bar_chart-books), so there doesn't have to be a conflict at all if only you had read the documentation.


All in all, this is not a bug, but a configuration error.
PHPCS follows the PSR12 specs to the letter (and I'm fairly certain this literal reading was verified by Greg when the ruleset was first created).
